### DISCUSSION TITLE
GNSS PMTK commands integration

### DISCUSSION THREAD

**Discussion by maxpowel** [zephyrproject-rtos/zephyr#75770] [Q&A]
Hello, I wrote a library that implements the MTK messages (https://docs.rs-online.com/3824/0900766b8147dbf6.pdf).
It is used by many quectel gnss devices so it could be useful for other people.

I don't know if this functionality is already implemented. I could not find it so if it already done please tell me where.

I currently use gnss-nmea-generic and it works great, but I want to extend it with these extra commands.
I want to do it in the "zephyr way" in the hope that this functionality could be in zephyr at some point.

I was studying this code https://github.com/zephyrproject-rtos/zephyr/tree/f320535f84db3b9bef8c1e975aa2fe61673c95c0/drivers/gnss and since these commands are proprietary it does not make sense to modify it but extend it.

Currently if I want to get the gnss data I do something like this:
```
GNSS_DATA_CALLBACK_DEFINE(DEVICE_DT_GET(DT_ALIAS(gnss)), gnss_data_cb);
```
Something similar for the satellites information. My idea is to add a new callback for the MTK messages:
```
static void pmtk_data_cb(const struct device *dev, const struct pmtk_data *data);
GNSS_PMTK_CALLBACK_DEFINE(DEVICE_DT_GET(DT_ALIAS(gnss)), pmtk_data_cb);
```
This callback is basically for ACKs and a little status information, nothing very extensive.
Looking at the zephyr code I know how to add a parser for new commands without any problem, so this is ok from my side.

The interesting part is that these gnss devices can receive commands (for constellation configuration, standby mode, custom features like geofence and much more).

Currently I have this code for sending the commands:
```
#define SUSPEND_CMD "$PMTK161,0*28\r\n"
...

const char *cmd = SUSPEND_CMD;
if (uart_tx(uart, cmd, strlen(cmd), SYS_FOREVER_MS)) {
       printk("Failed to send data over UART\n");
       return -1;
}
```
This works as expected and it can be easily wrapped into something to make it easier to use (like gnss->sleep()).
My first question is if is there any zephyr api that I can use for that (for sending command to the gnss device).

My second question is if this stuff could be integrated into zephyr or is it a very customized thing that should go as an external library.

If you think that this feature could be integrated into zephyr, I can create an pull request with my proposal and then work from it. 

Thank you!

**Comment by mxc42** [comment]
Did you end up getting any farther here? I'm in a similar spot.
